Source/FluentSpreadsheets/TableScaling/Scaler.cs

```diff
@@ -0,0 +1,127 @@
+    }
+
+    protected abstract void ValidateCustomization(IComponent component, IComponent customizedComponent);
+
+    protected abstract Axis GetAxis();
+
+    protected abstract IComponent MergeComponents(IEnumerable<IBaseComponent> componentSources);
+
+    protected abstract int SelectDimension(IComponent component);
+
+    private static void IterateEnumeratorNodeInDepth(IList<EnumeratorNode<IBaseComponent>> nodes)
+    {
+        for (var i = 0; i < nodes.Count; i++)
+        {
+            if (!nodes[i].HasValues)
+                continue;
+
+            while (nodes[i].Value is IComponentSource componentSource)
+            {
+                EnumeratorNode<IBaseComponent> node = nodes[i];
+                IEnumerator<IBaseComponent> enumerator = componentSource.GetEnumerator();
+                nodes[i] = new EnumeratorNode<IBaseComponent>(enumerator, node);
+                nodes[i].MoveNext();
+            }
+        }
+    }
+
+    private void ScaleEnumeratorNodesEdgeComponents(EnumeratorNode<IBaseComponent>[] nodes)
+    {
+        IEnumerable<int> dimensions = nodes
+            .Where(x => x.HasValues)
+            .Select(x => x.Value)
+            .OfType<IComponent>()
+            .Select(SelectDimension);
+
+        var dimension = LcmCounter.Count(dimensions);
+
+        foreach (EnumeratorNode<IBaseComponent> enumerator in nodes.Where(x => x.HasValues))
+        {
+            if (enumerator.Value is not IComponent component)
+                throw new InvalidOperationException("Max depth component source must be a component");
+
+            var factor = dimension / component.Size.Width;
+            var value = component;
+
+            if (factor is not 1)
+            {
+                value = component.ScaledBy(factor, GetAxis());
+            }
+
+            enumerator.Values.Add(value);
+        }
+    }
+
+    private void FlushDepletedEnumeratorNodes(IList<EnumeratorNode<IBaseComponent>> nodes)
+    {
+        for (var i = 0; i < nodes.Count; i++)
+        {
+            while (!nodes[i].HasValues && nodes[i].Next is not null)
+            {
+                EnumeratorNode<IBaseComponent> node = nodes[i];
+                node.Dispose();
+
+                if (node.Next?.Value is ICustomizerComponentSource customizerComponentSource)
+                {
+                    IComponent component = node.Values.Count switch
+                    {
+                        0 => Empty(),
+                        1 => (IComponent)node.Values[0],
+                        _ => MergeComponents(node.Values),
+                    };
+
+                    var customizedComponent = customizerComponentSource.Customize(component);
+                    ValidateCustomization(component, customizedComponent);
+                    nodes[i].Next!.Values.Add(customizedComponent);
+                }
+                else
+                {
+                    nodes[i].Next!.Values.AddRange(nodes[i].Values);
+                }
```

Comment:
Perhaps invert if and use `continue` and move some of the logic to another method, this one just looks bulky

Reply:
The thing with `continue` is: if you have a conditional logic that basically acts as a `Where` clause, you'll better off using it. If you have a logic that is more like "if this - do that, else - do that" it is clearer to use if-else statement. 

I can agree with you about block executed when condition is true being bulky, it is considerable to extract it to another function.